CARDS-1324: Unable to delete questions from questionnaire designer in some situations

sdumitriu opened the pull request:
All items in the questionnaire editor can be deleted, whether answered or not. Deleting a question will also delete all its answers from forms.

Possible problem: open a form in edit mode, in a different tab delete a question, fill in the answer and try to save the form. Does saving work, or report a problem?

sdumitriu committed:
CARDS-1324: Unable to delete questions from questionnaire designer in some situations
Fixed backend bug that prevented nested questions/sections from being deleted

marta- commented:
@veronikaslc In the questionnaire wizard, please rework the question/section delete button and ts confirmation to inform the user if there are any answers that will be deleted with that question/section.

E.g.:
* for unreferenced question/section: `Are you sure you want to delete question x?`

* for referenced: `Question X has 12 answers. Would you like to delete w and all items that reference it?`

Note: I see we're using a custom deletion workflow in the questionnaire wizard instead of the `<DeleteButton/>` component. Unless there are valid technical issues not to do so, we should probably use `DeleteButton` and its confirmation dialog, which already informs the users how many items are referencing the one being deleted.

Commit:
CARDS-1324: Unable to delete questions from questionnaire designer in some situations
Reworked delete button and ts confirmation to inform the user if there are any answers that will be deleted with that question/section

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/DeleteButton.jsx:
```diff
@@ -201,7 +201,7 @@ function DeleteButton(props) {
   variant: PropTypes.oneOf(["icon", "text", "extended"]), // "extended" means both icon and text
   label: PropTypes.string,
-  size: PropTypes.oneOf(["small", "medium"]),
+  size: PropTypes.oneOf(["small", "medium"])
```
Unnecessary change. I actually prefer `,` after the last item, since adding a new item requires changing only one line (the added item) instead of two (the added item plus a `,` at the end of the previous line).

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/QuestionnaireItemCard.jsx:
```diff
@@ -93,9 +92,13 @@ let QuestionnaireItemCard = (props) => {
+               entryPath={data["@path"]}
+               entryName={title || data.label || data.text || data["@name"]}
+               entryType={type}
+               warning={data["@referenced"] || false}
```
There's no `warning` prop in `DeleteButton`.
```suggestion
```

marta- reviewed (commented):
Looks good to me other than the two minor comments above.

marta- committed:
CARDS-1324: Unable to delete questions from questionnaire designer in some situations
Removed unnecessary changes

marta- reviewed (approved):
The wording of the delete confirmation dialog can be a bit awkward when deleting referenced sections: "_Section demographics_section is referenced in 2 answers and 1 other_", that 1 other being an `AnswerSection`. At the same time, mentioning the `AnswerSection` is probably too technical.

![image](https://user-images.githubusercontent.com/651980/131866894-34ce34f3-3cec-4c1e-919e-2bdbc09fab6a.png)

Regardless, this is a huge improvement from before, so it will do as is unless someone else has ideas of better wording right now.

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/DeleteServlet.java:
```diff
@@ -365,7 +367,11 @@ private String listReferrersFromTraversal()
                         break;
                     default:
-                        otherCount++;
+                        if ("cards/Answer".equals(n.getProperty("sling:resourceSuperType").getString())) {
+                            answerCount++;
```
Can't this be moved to a `case`?

marta- replied:
Let's also add a branch for `cards:AnswerSection`.

sdumitriu replied:
It cannot be moved to a case, since the switch is on `primaryNodeType`, which varies from answer to answer, but they have the same supertype. However, I would change this to a safer node type check, since not all nodes may have a `resourceSuperType property`:

```suggestion
                        if (n.isNodeType("cards:Answer")) {
                            answerCount++;
```

marta- replied:
Right, I see, thanks

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/DeleteServlet.java:
```diff
@@ -374,6 +380,7 @@ private String listReferrersFromTraversal()
             addNodesToResult(results, "subject", subjects);
             addNodesToResult(results, "subject type", subjectTypes);
             addNodesToResult(results, "questionnaire", questionnaires);
+            addNodesToResult(results, "answer", answerCount);
```
And let's add a label for it
```suggestion
            addNodesToResult(results, "answer", answerCount);
            addNodesToResult(results, "answer section", answerSectionCount);
```

marta- replied:
Right, I see, thanks

marta- reviewed (changes requested):
@veronikaslc It was decided to count answer sections and display them in the delete confirmation dialog.

Further improvements will be done separately as part of https://phenotips.atlassian.net/browse/CARDS-1379, out of scope for this PR.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/QuestionnaireItemCard.jsx:
```diff
@@ -30,7 +30,7 @@ import DeleteIcon from "@material-ui/icons/Delete";
 
 import EditDialog from "./EditDialog";
-import DeleteDialog from "./DeleteDialog";
+import DeleteButton from "../dataHomepage/DeleteButton.jsx";
```
This component seems to no longer belongs to the homepage only, it should be moved to a more generic place soon.

sdumitriu committed:
CARDS-1324: Unable to delete questions from questionnaire designer in some situations
Report the number of answer sections when deleting a referenced section

sdumitriu committed:
CARDS-1324: Unable to delete questions from questionnaire designer in some situations
Fixed NPE thrown sometimes when deleting answer sections

sdumitriu committed:
CARDS-1385: Too many questionnaires reported when deleting Patient subject type

sdumitriu commented:
Unintended consequence? After creating some forms, trying to delete Patient Information questionnaire reports that it is referenced by N forms, tens of answer sections, and hundreds of answers. This is too verbose. We should only report the top level nodes, without descending into them.

sdumitriu committed:
CARDS-1324: Unable to delete questions from questionnaire designer in some situations
Improvement: only list the highest affected node, not all the descendants, when forms/sections/answers will be affected by a deletion

sashaandjic commented:
Will this be improved in another PR by saying the section was referenced in 3 forms?
![image](https://user-images.githubusercontent.com/23243244/132078474-5e6a6eaa-43ec-49df-a763-2ea16abc6ca1.png)

sdumitriu commented:
> Will this be improved in another PR by saying the section was referenced in 3 forms?

Yes.

sashaandjic commented:
> > Will this be improved in another PR by saying the section was referenced in 3 forms?
> 
> Yes.

OK. I can mark this PR as tested then.

sdumitriu committed:
Merge pull request #768 from data-team-uhn/CARDS-1324-c

 Improvement: only list the highest affected node, not all the descendants, when forms/sections/answers will be affected by a deletion

sdumitriu committed:
Merge pull request #767 from data-team-uhn/CARDS-1385

CARDS-1385: Too many questionnaires reported when deleting Patient subject type
